src/internal/components/token-list/index.tsx

```diff
@@ -0,0 +1,69 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import clsx from 'clsx';
+import React, { useState } from 'react';
+import styles from './styles.css.js';
+import { SelectToggle } from './select-toggle';
+import { I18nStrings, ItemAttributes } from './interfaces';
+import { useUniqueId } from '../../hooks/use-unique-id';
+
+interface TokenListProps<Item> {
+  alignment: 'vertical' | 'horizontal';
+  toggleAlignment?: 'vertical' | 'horizontal';
+  items: readonly Item[];
+  limit?: number;
+  renderItem: (item: Item, itemIndex: number) => React.ReactNode;
+  getItemAttributes: (item: Item, itemIndex: number) => ItemAttributes;
+  i18nStrings?: I18nStrings;
+}
+
+export function TokenList<Item>({
+  alignment = 'vertical',
+  toggleAlignment = 'vertical',
+  items,
+  limit,
+  renderItem,
+  getItemAttributes,
+  i18nStrings,
+}: TokenListProps<Item>) {
+  const controlId = useUniqueId();
+
+  const [expanded, setExpanded] = useState(false);
+  const hasItems = items.length > 0;
+  const hasHiddenItems = hasItems && limit !== undefined && items.length > limit;
+  const slicedItems = hasHiddenItems && !expanded ? items.slice(0, limit) : items;
+
+  return (
+    <div className={clsx(styles.root, styles[`root-${toggleAlignment}`])}>
+      <ul id={controlId} className={clsx(styles.list, styles[`list-${alignment}`])}>
+        {slicedItems.map((item, itemIndex) => {
+          const { ariaLabel, disabled } = getItemAttributes(item, itemIndex);
+          return (
+            <li
+              key={itemIndex}
+              aria-label={ariaLabel}
```

Comment:
If this label is required it means that token content is never used. Is this expected?

Reply:
Checked in Chrome devtools, the content of token group is marked as "Ignored" in AOM. This does not look good at all

<img width="1225" alt="image" src="https://user-images.githubusercontent.com/812240/228215056-ed38db84-9427-4ca0-adee-7c493796cf3f.png">

Reply:
(this is likely due to the `role="presentation"` I think https://github.com/cloudscape-design/components/pull/876/files#r1149456787)

Reply:
The role="presentation" makes sure the internally used option is not treated as a group. Otherwise there are two groups - the token itself and the option with the last one being not complete as it does not include the dismiss button.

The testing with screen-readers does not reveal any problems yet - all content is announced and properly grouped. Could you suggest a testing scenario that confirms the change causes content not being accessible?

Reply:
Extra content in tokens (description and tags) is concealed when aria-label is used: https://d21d5uik3ws71m.cloudfront.net/components/df9ba4bdca99f9addd3f836354a0b24238cad39c/index.html#/light/token-group/permutations

> The testing with screen-readers does not reveal any problems yet

Did the missing labels revealed any problems? Why not to start with a simple option?

Reply:
@pan-kot still see the same outcome in the latest revision.

I do not think any parent role changes the behavior or aria label. It is an ultimate feature – when you apply the label, all inner content gets ignored.

Still curious to know reasons why you added the label in the first place

Reply:
Also, [quoting MDN](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label)

> Note: aria-label is intended for use on interactive elements, or elements made to be interactive via other ARIA declarations

This does not sound suitable for this use-case

Reply:
The ARIA label was added based on the feedback from a11y - it maintains the context so that if you navigate with Tab to the dismiss button the group name described with ARIA label will be announced. Thanks for sharing the concerns though, I'll double-check the approach.

Reply:
When I open in my inspector, I see the expanded content, but only when focusing one particular node. When going up or down, the tree shows "ignored"

https://user-images.githubusercontent.com/812240/228520387-3e45c95e-de8f-48b0-a0b2-b2425a6700cc.mov

Reply:
I also double checked what it means in Voice Over interpretation. Currently it announces the title twice (label causes the duplication, I am guessing).

> The ARIA label was added based on the feedback from a11y - it maintains the context so that if you navigate with Tab to the dismiss button the group name described with ARIA label will be announced

The `dismissLabel` is unique for each item. What extra context missing, I do not understand.

Reply:
Here is the feedback I got:

"We want a label on the item (which is a group) so that when a user reaches a control inside of it they know which one they are affecting. The user may not have reached that control through linear or tab navigation, it could be the result of navigating by button, for example."

Of course, as it appears the addition of the aria-label causes more issues the decision needs to be re-evaluated. I'll check it

Reply:
Turns out the "Ignored" property assigned in Chrome (but not e.g. Firefox) is not the same as hidden. It also applies e.g. when an element has role "presentation" or "none" set and means that the element's semantics not exposed to the a11y tree so the element e.g. can be excluded from the list this way. However, it does not prevent screen-readers to navigate inside. W/o "aria-label" set and "Ignored" assigned the option title is announced when entering the group. When the node is ignored its contents is not announced when entering the group and is being replaced by the ARIA-label contents. However, having an aria label set has benefits and now when the group element is navigated by clicking or using Tab navigation the group name is announced.

You can see there are plenty of "Ignored" nodes in e.g. our multselect test page.

<img width="1413" alt="Screenshot 2023-03-31 at 15 40 56" src="https://user-images.githubusercontent.com/20790937/229135967-287b82cf-97a8-4810-a981-4af6df8726a2.png">